Expand Navigation docs (#1061)
solidjs/solid-docs: src/routes/solid-router/concepts/navigation.mdx

title: "Navigation"
---

In Solid Router, standard HTML [`<a>` elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) trigger [soft navigation](/solid-router/reference/components/a#soft-navigation).
Solid Router also provides these options for navigating between routes:

- The [`<A>` component](/solid-router/reference/components/a).
- The [`useNavigate` primitive](/solid-router/reference/primitives/use-navigate).

> - The [`redirect` function](/solid-router/reference/response-helpers/redirect).

Reviewer: I might've lied - do you mind putting the anchor text in a callout? It feels a bit misplaced in the first sentence but having it in the intro section does makes sense

Author: Done. Still, it feels out of place. Maybe we can add more context:

> Unlike other routers, in Solid Router you don't have to rely on a specific component (for example, `<Link>`) for client-side navigation. Solid Router registers a click listener that captures click events in the `<a>` elements and triggers a client-side navigation. 

Something like this feels less out of place IMO and also provides more information on how Solid Router works.

> This page provides an overview of these options.

Reviewer: ```suggestion
```

title: "Navigation"
---

> <Callout type="info" title="Note">

Reviewer: I don't particularly like the location of this callout since it feels a lot more out of place. I think I'd rather have it located within the introduction, like I believe was talked about before. 

Like this:

```
When using Solid Router, you can use the standard standard HTML [`<a>` elements](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a), which triggers [soft navigation](/solid-router/reference/components/a#soft-navigation). 
In addition to using this, Solid Router offers other options for navigating between routes:

// list here
```

title: "Navigation"
---

> SolidStart provides several ways to manage navigation between routes:

Reviewer: ```suggestion
In a SolidStart application, there are several ways to navigate between routes outside of the [native `<a>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a).
This includes the following options provided by [Solid Router](/solid-router):
```

Author: This is a mistake on my part. It should be "Solid Router" not "SolidStart". Maybe we should add a similar page in the SolidStart docs though.

I also don't like the casual mention of the `<a>` element. Yeah, It's not provided by Solid Router, but its behaviour is changed by Solid Router. I prefer to have a dedicated section about it to explain that it triggers a client-side navigation in detail.

Reviewer: See my other comment.

- Using the native [`<a>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a).
- Using the [`<A>` component](/solid-router/reference/components/a).
- Using the [`useNavigate` primitive](/solid-router/reference/primitives/use-navigate).
- Using the [`redirect` function](/solid-router/reference/response-helpers/redirect).

> This page provides an overview of these options.

Reviewer: ```suggestion
- [`<A>` component](/solid-router/reference/components/a).
- [`useNavigate` primitive](/solid-router/reference/primitives/use-navigate).
- [`redirect` function](/solid-router/reference/response-helpers/redirect).
```

Author: As explained above, keeping the `<a>` element makes it clear that `<a>` elements have a special behaviour in Solid Router. Otherwise, they might be seen as a regular `<a>` elements that perform a server-side navigation.

Reviewer: It's a fine line between teaching people the basics (such as basic JS and HTML) and just informing them on Solid. Upon further thought, I think that it's more appropriate to add the soft navigation part as a callout within the `<A>` component section (or possible above it) instead of having an area showing people how to use it.

Author: Does the new change I pushed address this?

Reviewer: yes! thank you

## `<a>` element

> The standard HTML `<a>` element is the most basic way to create links.

Reviewer: ```suggestion
The standard HTML `<a>` element is the universal way to create links in an application.
```

> In SolidStart, clicking an `<a>` element triggers [soft navigation](/solid-router/reference/components/a#soft-navigation).

Reviewer: ```suggestion
In SolidStart, the `<a>` element triggers [soft navigation](/solid-router/reference/components/a#soft-navigation).

```

```tsx
function Navbar() {
	return (
		<nav>
			<a href="/">Home</a>
			<a href="/login">Login</a>
		</nav>
	);
}
```

## `<A>` component

The `<A>` component extends the native `<a>` element.

> It automatically handles the router's base URL and supports relative paths.

Reviewer: ```suggestion
The `<A>` component extends the native `<a>` element by automatically applying the base URL path and, additionally, supports relative paths.
```

```tsx
import { A } from "@solidjs/router";

function DashboardPage() {
	return (
		<main>
			{/* This is a relative path that, from /dashboard, links to /dashboard/users */}
			<A href="users">Users</A>

> 		</main>

Reviewer: ```suggestion
		<nav>
		        <A href="/">Home</A>
			<A href="/users">Users</A>
		</nav>

```

Author: I like to keep the relative path example because I didn't find an example anywhere else.

If it's not very clear that the `href` prop of `<A>` is exactly like `<a>`, I can add a regular link too.

	);
}
```

<Callout type="info" title="Note">

`<A>` can only be used within a component that is rendered by a [`<Route>`](/solid-router/reference/components/route).

</Callout>

> See the [`<A>` API reference](/solid-router/reference/components/a) for more information.

Reviewer: Callout is something that should be included in the router section instead of SolidStart.

```suggestion
See the [`<A>` API reference](/solid-router/reference/components/a) in the Solid Router section for more information.
```

Author: Again, my mistake. This page is in the Solid Router docs.

### Styling

The `<A>` component provides `activeClass` and `inactiveClass` props to style links based on their active state.
These props apply corresponding CSS classes to the link.

> If these props are omitted, the classes `active` and `inactive` are used by default.

Reviewer: ```suggestion
The `<A>` component allows you to style links based on their active state using the `activeClass` and `inactiveClass` props. 
When provided, these props apply the corresponding CSS classes to the link. 
If omitted, the default classes `active` and `inactive` are used.

```

> A link is considered active if the current route matches the link's `href` or is a descendant of it.
> For example, a link with `href="/dashboard"` is active when the current route is `/dashboard`, `/dashboard/users`, or `/dashboard/users/1/profile`.

Reviewer: ```suggestion
By default, a link is considered active when the current route matches the `href` or any of its descendant routes.
For example, if the route `/users` is specified, it would also match `/users`, `/users/1`, `/users/1/posts`, and any other nested routes.
```

Author: > if the route `/users` is specified

It's not clear that _where_ `/users` is specified.

> The `end` prop modifies this behavior.
> `end={true}` requires an exact match between the link's `href` and the current route for the link to be considered active.
> This is particularly useful for root route links (`href="/"`), which would otherwise match all routes.

Reviewer: ```suggestion
To match an exact route, the prop `end` can be used.
When `true`,  it ensures that the link is only considered active if the `href` exactly matches the current route. 
This is useful for root route links (href="/"), which might otherwise match all routes.
```

```tsx
import { A } from "@solidjs/router";

function Navbar() {
	return (
		<nav>
			<A href="/" end={true}>
				Home
			</A>
			<A
				href="/login"
				activeClass="text-blue-900"
				inactiveClass="text-blue-500"
			>
				Login
			</A>
		</nav>
	);
}
```

## `useNavigate` primitive

The `useNavigate` primitive allows for programmatically navigating to a specified route.

```tsx
import { useNavigate } from "@solidjs/router";

function LoginPage() {

> 	const navigate = useNavigate();
> 
> 	return (
> 		<button
> 			onClick={() => {
> 				// Perform login logic here ...
> 				navigate("/dashboard", { replace: true });
> 			}}
> 		>
> 			Login
> 		</button>
> 	);

Reviewer: ```suggestion
	const navigate = useNavigate();

	return (
		<button
			onClick={() => {
		                // Login logic
				navigate("/dashboard", { replace: true });
			}}
		>
			Login
		</button>
	);

```

}
```

> This example redirects the user to `/dashboard` after login.
> The `replace: true` option effectively removes the login page from the browser's history stack by replacing it with the `/dashboard` route, preventing the user from returning to it with the back button.

Reviewer: ```suggestion
This example redirects the user to `/dashboard` after login. 
By using `replace: true`, the login page is removed from the browser's history stack and replaced with the `/dashboard` route. 
This prevents the user from navigating back to the login page using the back button.
```

> <Callout type="info" title="Note">
> 
> `useNavigate` can only be used within a component that is rendered by a `<Route>`.
> 
> </Callout>

Reviewer: should be added to `useNavigate` not here
```suggestion
```

See the [`useNavigate` API reference](/solid-router/reference/primitives/use-navigate) for more information.

## `redirect` function

> The `redirect` function returns a [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object that allows navigation to a specified route within a [query](/solid-router/reference/data-apis/query) or [action](/solid-router/reference/data-apis/action).

Reviewer: ```suggestion
The `redirect` function returns a [`Response` object](https://developer.mozilla.org/en-US/docs/Web/API/Response), which enables navigation to a specified route within [query](/solid-router/reference/data-apis/query) or [action](/solid-router/reference/data-apis/action).
```
